include/rapidcheck/state/Commands.hpp

```diff
@@ -1,9 +1,109 @@
 #pragma once
 
 #include "rapidcheck/detail/Results.h"
+#include "rapidcheck/state/gen/ParallelCommands.hpp"
+#include "rapidcheck/Assertions.h"
+#include <thread>
+#include <mutex>
+#include "rapidcheck/detail/Barrier.h"
 
 namespace rc {
 namespace state {
+namespace detail {
+
+
+template <typename Model, typename Cmd>
+struct CommandResult
+{
+  CommandResult(const std::shared_ptr<const Cmd>& command,
+                std::function<void(const Model&)>&& verifyFunc)
+    : command(command)
+    , verifyFunc(std::move(verifyFunc))
+  {
+  }
+
+  std::shared_ptr<const Cmd> command;
+  std::function<void(const Model&)> verifyFunc;
+};
+
+template <typename Model, typename Cmd>
+struct ParallelExecutionResult {
+  std::vector<CommandResult<Model, Cmd>> prefix;
+  std::vector<CommandResult<Model, Cmd>> left;
+  std::vector<CommandResult<Model, Cmd>> right;
+};
+
+template <typename Model, typename Cmd>
+bool hasValidInterleaving(const std::vector<CommandResult<Model, Cmd>> &left,
+                          const std::vector<CommandResult<Model, Cmd>> &right,
+                          std::size_t leftIndex,
+                          std::size_t rightIndex,
+                          const Model &state) {
+  const auto hasLeft = leftIndex < left.size();
+  const auto hasRight = rightIndex < right.size();
+
+  // Base case, no more commands to verify
+  if (!hasRight && !hasLeft) {
+    return true;
+  }
+
+  if (hasLeft) {
+    try {
+      auto preState = state;
```

Comment:
This is well intended but `state` is potentially a very expensive object to copy and there is really no need to do it here since we're not modifying it.